File: client/src/ecs/cluster.js

import delay from './delay';
export const CLUSTER_ACTIVE = "ACTIVE";
export const CLUSTER_INACTIVE = "INACTIVE";

const clusters = [
  {
    activeServiceCount: 0,
    clusterName: "minecraft",
    pendingTasksCount: 0,
    registeredContainerInstanceCount: 0,
    runningTasksCount: 0,
    status: CLUSTER_INACTIVE
  },
  {
    activeServiceCount: 0,
    clusterName: "craft-staging",
    pendingTasksCount: 2,
    registeredContainerInstanceCount: 2,
    runningTasksCount: 8,
    status: CLUSTER_ACTIVE
  },
  {
    activeServiceCount: 0,
    clusterName: "minecraft-bungee",
    pendingTasksCount: 0,
    registeredContainerInstanceCount: 0,
    runningTasksCount: 0,
    status: CLUSTER_ACTIVE
  },
  {
    activeServiceCount: 0,
    clusterName: "craft-production",
    pendingTasksCount: 0,
    registeredContainerInstanceCount: 2,
    runningTasksCount: 9,
    status: CLUSTER_ACTIVE
  },
  {
    activeServiceCount: 1,
    clusterName: "log-production",
    pendingTasksCount: 0,
    registeredContainerInstanceCount: 1,
    runningTasksCount: 1,
    status: CLUSTER_ACTIVE
  }
];


export default class Cluster {
  static getClusters() {
    return new Promise((resolve) => {
      setTimeout( () => {
        resolve(Object.assign([], clusters));
      }, delay);
    });
  }

  static byName(clusters) {
    return clusters.sort((a,b) => {
      const an = a.clusterName;
      const bn = b.clusterName;
      return an < bn ? -1 : (bn < an ? 1  : 0);
    });
  }
}

File: server/cluster.go
package server

import (
  "fmt"
  // "encoding/json"
  "net/http"

  "github.com/aws/aws-sdk-go/aws"
  "github.com/aws/aws-sdk-go/service/ecs"
  "github.com/aws/aws-sdk-go/private/protocol/json/jsonutil"
  // jwt "github.com/dgrijalva/jwt-go"
  "github.com/Sirupsen/logrus"
  "github.com/jdrivas/awslib"
)

var TestClusters = []ecs.Cluster {
  ecs.Cluster{
    ActiveServicesCount: aws.Int64(0),
    ClusterName: aws.String("minecraft"),
    PendingTasksCount: aws.Int64(0),
    RegisteredContainerInstancesCount: aws.Int64(0),
    RunningTasksCount: aws.Int64(0),
    Status: aws.String("INACTIVE"),
  },
  ecs.Cluster{
    ActiveServicesCount: aws.Int64(0),
    ClusterName: aws.String("craft-staging"),
    PendingTasksCount: aws.Int64(2),
    RegisteredContainerInstancesCount: aws.Int64(2),
    RunningTasksCount: aws.Int64(8),
    Status: aws.String("ACTIVE"),
  },
}

func ClusterController(w http.ResponseWriter, r *http.Request) {
  f := logrus.Fields{"controller": "ClusterController"}

  sess, err := getAWSSession(r)
  if err != nil {
    log.Error(f, "Fail to find appropriate AWS Session", err)
    http.Error(w, fmt.Sprintf("Failed to find appropriate AWS Session: %s", err), http.StatusFailedDependency )
    return
  }

  // clusters, err := awslib.GetAllClusterDescriptions(awsSession)
  clusters, err := awslib.GetAllClusterDescriptions(sess)
  if err != nil {
    log.Error(f, "Failed to get current set of clusters from AWS", err)
    http.Error(w, fmt.Sprintf("Failed to obtain clusters from AWS: %s", err), http.StatusFailedDependency)
    return
  }

  f["numberOfClusters"] = len(clusters)
  log.Debug(f, "Got clusters from Amazon")

  clusterJson, err  := jsonutil.BuildJSON(clusters)
  if err != nil {
    log.Error(f, "Failed to marshall JSON on clusters.", err)
    http.Error(w, "Failed to marshall JSON for response.", http.StatusInternalServerError)
    return
  }

  _, err = w.Write(clusterJson)
  if err != nil {
    log.Error(f, "Failed to write JSON response", err)
  } else {
    f["json"] = string(clusterJson)
    log.Debug(f, "Sent repsonse.")
  }
}
